pkg/autoscaler/autoscaler/util.go

```diff
@@ -0,0 +1,133 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+)
+
+// checkStsAutoScalingPrerequisites would check the sts status to ensure wouldn't happen during
+// upgrading, scaling
+func checkStsAutoScalingPrerequisites(set *appsv1.StatefulSet) bool {
+	if operatorUtils.IsStatefulSetUpgrading(set) {
+		return false
+	}
+	if operatorUtils.IsStatefulSetScaling(set) {
+		return false
+	}
+	return true
+}
+
+// checkStsAutoScalingInterval would check whether there is enough interval duration between every two auto-scaling
+func checkStsAutoScalingInterval(tc *v1alpha1.TidbCluster, intervalSeconds int32, memberType v1alpha1.MemberType) (bool, error) {
+	if tc.Annotations == nil {
+		tc.Annotations = map[string]string{}
+	}
+	lastAutoScalingTimestamp, existed := tc.Annotations[label.AnnTiDBLastAutoScalingTimestamp]
+	if memberType == v1alpha1.TiKVMemberType {
+		lastAutoScalingTimestamp, existed = tc.Annotations[label.AnnTiKVLastAutoScalingTimestamp]
+	}
+	if !existed {
+		return true, nil
+	}
+	t, err := strconv.ParseInt(lastAutoScalingTimestamp, 10, 64)
+	if err != nil {
+		return false, err
+	}
+	if intervalSeconds > int32(time.Now().Sub(time.Unix(t, 0)).Seconds()) {
+		return false, nil
+	}
+	return true, nil
+}
+
+// checkAutoScalingPrerequisites would check the tidbcluster status to ensure the autoscaling would'n happen during
+// upgrading, scaling and syncing
+func checkAutoScalingPrerequisites(tc *v1alpha1.TidbCluster, sts *appsv1.StatefulSet, memberType v1alpha1.MemberType) bool {
+	if !checkStsAutoScalingPrerequisites(sts) {
+		return false
+	}
+	if memberType == v1alpha1.TiDBMemberType {
+		if tc.Status.TiDB.Phase != v1alpha1.NormalPhase {
+			return false
+		}
+	} else if memberType == v1alpha1.TiKVMemberType {
+		if !tc.Status.TiKV.Synced {
+			return false
+		}
+		if tc.Status.TiKV.Phase != v1alpha1.NormalPhase {
+			return false
+		}
+	} else {
+		// Unknown MemberType
+		return false
+	}
+	return true
+}
+
+// limitTargetReplicas would limit the calculated target replicas to ensure the min/max Replicas
+func limitTargetReplicas(targetReplicas int32, tac *v1alpha1.TidbClusterAutoScaler, memberType v1alpha1.MemberType) int32 {
+	if targetReplicas > tac.Spec.TiKV.MaxReplicas {
+		targetReplicas = tac.Spec.TiKV.MaxReplicas
+	} else if targetReplicas < *tac.Spec.TiKV.MinReplicas {
+		targetReplicas = *tac.Spec.TiKV.MinReplicas
+	}
+	if memberType == v1alpha1.TiDBMemberType {
+		if targetReplicas > tac.Spec.TiDB.MaxReplicas {
+			targetReplicas = tac.Spec.TiDB.MaxReplicas
+		} else if targetReplicas < *tac.Spec.TiDB.MinReplicas {
+			targetReplicas = *tac.Spec.TiDB.MinReplicas
+		}
+	}
+	return targetReplicas
+}
+
+// If the minReplicas not set, the default value would be 1
+// If the Metrics not set, the default metric will be set to 80% average CPU utilization.
+// defaultTAC would default the omitted value
+func defaultTAC(tac *v1alpha1.TidbClusterAutoScaler) {
+	defaultMetricSpec := autoscalingv2beta2.MetricSpec{
```

Comment:
Make this a variable out of the function?

Reply:
updated.